integration/upload_test.go

```diff
@@ -0,0 +1,257 @@
+// https://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+
+//go:build integration
+
+package integration
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/fs"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"github.com/caarlos0/env/v6"
+	"github.com/eclipse-kanto/file-upload/client"
+	"github.com/eclipse-kanto/file-upload/uploaders"
+	"github.com/eclipse-kanto/kanto/integration/util"
+	"github.com/eclipse/ditto-clients-golang/protocol"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+)
+
+type uploadRequest struct {
+	CorrelationID string            `json:"correlationId"`
+	Options       map[string]string `json:"options"`
+}
+
+type uploadStatus struct {
+	State    string `json:"state"`
+	Progress int    `json:"progress"`
+}
+
+func (suite *fileUploadSuite) SetupSuite() {
+	suite.Setup(suite.T())
+
+	opts := env.Options{RequiredIfNoDef: false}
+	suite.uploadCfg = uploadTestConfig{}
+	require.NoError(suite.T(), env.Parse(&suite.uploadCfg, opts), "failed to process upload environment variables")
+	suite.T().Logf("upload test configuration - %v", suite.uploadCfg)
+	suite.checkUploadDir()
+
+	fmt.Println(suite.ThingCfg)
+	suite.thingURL = util.GetThingURL(suite.Cfg.DigitalTwinAPIAddress, suite.ThingCfg.DeviceID)
+	suite.featureURL = util.GetFeatureURL(suite.thingURL, featureID)
+}
+
+func (suite *fileUploadSuite) TearDownSuite() {
+	suite.TearDown()
+}
+
+func TestFileUploadHTTP(t *testing.T) {
+	suite.Run(t, new(httpFileUploadSuite))
+}
+
+func TestFileUploadAzure(t *testing.T) {
+	suite.Run(t, new(azureFileUploadSuite))
+}
+
+func TestFileUploadAWS(t *testing.T) {
+	suite.Run(t, new(awsFileUploadSuite))
+}
+
+func (suite *httpFileUploadSuite) TestFileUpload() {
+	if len(suite.uploadCfg.HTTPServer) == 0 {
+		suite.T().Fatal("http server must be set")
+	}
+	upload := newHTTPUpload(suite.T(), suite.uploadCfg.HTTPServer)
+	suite.testUpload(upload)
+}
+
+func (suite *azureFileUploadSuite) TestFileUpload() {
+	creds, err := uploaders.GetAzureTestCredentials()
+	require.NoError(suite.T(), err, "please set azure environment variables.")
+	options, err := uploaders.GetAzureTestOptions(creds)
+	require.NoError(suite.T(), err, "error getting azure test options")
+	upload := newAzureUpload(suite.T(), options)
+	suite.testUpload(upload)
+}
+
+func (suite *awsFileUploadSuite) TestFileUpload() {
+	creds, err := uploaders.GetAWSTestCredentials()
+	require.NoError(suite.T(), err, "please set aws environment variables.")
+	options := uploaders.GetAWSTestOptions(creds)
+	upload, err := newAWSUpload(suite.T(), options)
+	require.NoError(suite.T(), err, "error creating AWS client")
+	suite.testUpload(upload)
+}
+
+func (suite *fileUploadSuite) triggerUploads() map[string]string {
+	topicRequest := util.GetLiveMessageTopic(suite.ThingCfg.DeviceID, protocol.ActionRequest)
```

Comment:
I am curious whether it's ok to reuse this predefined request action. The issue is that it's a part of the search: https://www.eclipse.org/ditto/protocol-specification-topic.html#search-criterion-actions. I guess that upload specific request action has to be defined.